README.md

```diff
@@ -4,222 +4,1008 @@
+
+Let's now create our new command:
+
+```d
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    File file;
+
+    void onExecute()
+    {
+        import std.stdio : writeln;
+
+        foreach(lineInFile; this.file.byLine())
+            writeln(lineInFile);
+    }
+}
+```
+
+The most important thing of note here is, notice how the `file` variable has the type `File`, and recall that our arg binder's `output` parameter also has the type `File`? This tells the arg binder that it has a function to convert the user's provided string into a `File` for us.
+
+Our `onExecute` function is nothing overly special, it just displays the file line by line.
+
+Test time. Let's make it show the contents of our `dub.json` file, which is within the root of our project:
+
+```bash
+$> ./mytool cat ./dub.json
+{
+    "authors": [
+            "Sealab"
+    ],
+    "copyright": "Copyright ┬® 2020, Sealab",
+    "dependencies": {
+            "jcli": "~>0.10.0"
+    },
+    "description": "A minimal D application.",
+    "license": "proprietary",
+    "name": "mytool"
+}
+Program exited with status code 0
+
+# And just for good measure, let's see what happens if the file doesn't exist
+$> ./mytool cat non-existing-file
+std.exception.ErrnoException@std\stdio.d(428): Cannot open file `non-existing-file' in mode `r' (No such file or directory)
+```
+
+Very simple. Very useful.
+
+## User Defined argument validation
+
+It's cool and all being able to very easily create arg binders, but sometimes commands will need validation logic involved.
+
+For example, some commands might only want files with a `.json` extention, while others may not care about extentions. So putting this logic into the arg binder itself isn't overly wise.
+
+Some arguments may need validation on the pre-arg-binded string, whereas others may need validation on the post-arg-binded value. Some may need both!
+
+JCLI handles all of this via argument validators.
+
+Let's start off with the first example, making sure the user only passes in files with a `.json` extention, and apply it to our `cat` command. Code first, explanation after:
+
+```d
+struct HasExtention
+{
+    string wantedExtention;
+
+    bool onPreValidate(string arg, ref string errorIfFalse)
+    {
+        import std.algorithm : endsWith;
+
+        // Errors only display if we return false.
+        errorIfFalse = "Expected file to have extention of "~this.wantedExtention;
+
+        return arg.endsWith(this.wantedExtention);
+    }
+}
+
+@Command("cat", "Displays the contents of a file.")
+struct CatCommand
+{
+    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
+    @HasExtention(".json")
```

Comment:
Possibly...

It's something I'm going to need to think over, as I'm unsure about the benefits of something like that.

It's not like you can even reuse functions in the different validation stages, since pre validation and post validation target different types, and I don't see room for any other validation stages.

The only benefit I'm seeing would potentially be clarity, but even then a properly named validator in the current style may be all you need.

I'm too unsure about it to really consider changing over to that.